### DISCUSSION TITLE
@adonisjs/ally Twitter compatibility

### DISCUSSION THREAD

**Discussion by DraftProducts** [adonisjs/core#3498] [Help] [answered]
Hi guys 👋 
I've tried to use [`@adonisjs/ally`](https://github.com/adonisjs/ally) which most of the Drivers, but I issued a problem with Twitter auth.
Please note that Twitter is the only one Driver witch use OAuth1 in ally library and I try to get `redirectUrl()` directly without going throw redirect() method.

Each time when I try to connect with Twitter, I get an empty authorization url : `https://api.twitter.com/oauth/authenticate`.

Here is the code that I use to auth with Twitter :
```ts
public async twitterRedirection({ ally }: HttpContextContract) {
  return ally.use('twitter').redirectUrl();
}

public async twitterCallback({ auth, response, ally }: HttpContextContract) {
  const twitter = ally.use('twitter')

  // Errors verifications

  const twitterUser = await twitter.user()

  console.log(twitterUser)

  const user = await User.updateOrCreate(
    {
      email: twitterUser.email!,
    },
    {
      username: twitterUser.name,
      loginSource: SocialAccounts.TWITTER,
    }
  )

  const { token } = await auth.use('user').generate(user)

  return response.ok({ token })
}
```

Is there a way to use [`@adonisjs/ally`](https://github.com/adonisjs/ally) with a backend API based ?

Is there any information that I forgot, that I didn't understand ?

Thank you for your help !

**Comment by DraftProducts** [comment, marked as answer]
Okay, it looks like that return to frontend the redirect uri to make the redirection self is only possible with Oauth2.

**Reply by Akpagni5547** [reply]
how to specify oauth2 with ally.use('twitter')

**Reply by Barbapapazes** [reply]
You can't actually. You must write another driver or update the actual one

**Reply by Akpagni5547** [reply]
oh thanks. Can you share me something to do this ?

**Reply by Barbapapazes** [reply]
Yes, this is a boilerplate, https://docs.adonisjs.com/guides/auth/social#adding-custom-drivers

Or here the driver https://github.com/adonisjs/ally/blob/develop/src/Drivers/Twitter/index.ts

**Reply by Akpagni5547** [reply]
thanks 👍

**Comment by darshithedpara** [comment]
@Akpagni5547  Have you found any solution for twitter with OAuth2 ?

**Reply by naveenbasyal** [reply]
Have u fnid any solution?
